src/openrct2-ui/windows/Viewport.cpp

```diff
@@ -10,192 +10,217 @@
+    MakeWidget({WW - 25, 41}, VIEWPORT_BUTTON,   WindowWidgetType::FlatBtn,  WindowColour::Primary  , SPR_G2_ZOOM_OUT, STR_ZOOM_OUT_TIP      ), // zoom out
+    MakeWidget({WW - 25, 65}, VIEWPORT_BUTTON,   WindowWidgetType::FlatBtn,  WindowColour::Primary  , SPR_LOCATE,      STR_LOCATE_SUBJECT_TIP), // locate
     { WIDGETS_END },
 };
 
-static void window_viewport_mouseup(rct_window *w, rct_widgetindex widgetIndex);
-static void window_viewport_resize(rct_window *w);
-static void window_viewport_update(rct_window *w);
-static void window_viewport_invalidate(rct_window *w);
-static void window_viewport_paint(rct_window *w, rct_drawpixelinfo *dpi);
+// clang-format on
 
-static rct_window_event_list window_viewport_events([](auto& events)
+class ViewportWindow final : public Window
 {
-    events.mouse_up = &window_viewport_mouseup;
-    events.resize = &window_viewport_resize;
-    events.update = &window_viewport_update;
-    events.invalidate = &window_viewport_invalidate;
-    events.paint = &window_viewport_paint;
-});
-// clang-format on
+private:
+    int32_t _viewportWindowNumber = 1;
 
-static int32_t _viewportNumber = 1;
+    void get_free_viewport_number(rct_window* w)
+    {
+        if (w != nullptr && w->classification == WC_VIEWPORT)
+        {
+            if (w->number >= _viewportWindowNumber)
+                _viewportWindowNumber = w->number + 1;
+        }
+    }
 
-/**
- * Creates a custom viewport window.
- */
-rct_window* window_viewport_open()
-{
-    rct_window* w = WindowCreateAutoPos(WW, WH, &window_viewport_events, WC_VIEWPORT, WF_RESIZABLE);
-    w->widgets = window_viewport_widgets;
-    w->enabled_widgets = (1ULL << WIDX_CLOSE) | (1ULL << WIDX_ZOOM_IN) | (1ULL << WIDX_ZOOM_OUT) | (1ULL << WIDX_LOCATE);
-    w->number = _viewportNumber++;
-
-    // Create viewport
-    viewport_create(w, w->windowPos, w->width, w->height, 0, TileCoordsXYZ(128, 128, 0).ToCoordsXYZ(), 1, SPRITE_INDEX_NULL);
-    rct_window* mainWindow = window_get_main();
-    if (mainWindow != nullptr)
+public:
+    void OnOpen() override
     {
-        rct_viewport* mainViewport = mainWindow->viewport;
-        int32_t x = mainViewport->viewPos.x + (mainViewport->view_width / 2);
-        int32_t y = mainViewport->viewPos.y + (mainViewport->view_height / 2);
-        w->savedViewPos = { x - (w->viewport->view_width / 2), y - (w->viewport->view_height / 2) };
+        window_visit_each([&](rct_window* w) { get_free_viewport_number(w); });
+        number = _viewportWindowNumber;
+
+        widgets = window_viewport_widgets;
+        enabled_widgets = (1ULL << WIDX_CLOSE) | (1ULL << WIDX_ZOOM_IN) | (1ULL << WIDX_ZOOM_OUT) | (1ULL << WIDX_LOCATE);
+
+        // Create viewport
+        viewport_create(this, windowPos, width, height, 0, TileCoordsXYZ(128, 128, 0).ToCoordsXYZ(), 1, SPRITE_INDEX_NULL);
```

Comment:
`viewport_create` can fail, in which case, `viewport` will be `nullptr`. You should there for check `viewport` is not `nullptr` before doing anything with it. This obviously improves on the original code which didn't check.

Reply:
Added a null check and it will Close the window and display an error tooltip to say something went wrong.